src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ConstructorsDeclarationGroupingCheck.java

```diff
@@ -0,0 +1,104 @@
+///////////////////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code and other text files for adherence to a set of rules.
+// Copyright (C) 2001-2024 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+///////////////////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.coding;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import com.puppycrawl.tools.checkstyle.FileStatefulCheck;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+
+/**
+ * <p>
+ * Checks that overloaded constructors are grouped together.
+ * If there is anything between overloaded constructors ( expect comments )
+ * then this check will give an error.
+ * </p>
+ * <p>
+ * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
+ * </p>
+ * <p>
+ * Violation Message Keys:
+ * </p>
+ * <ul>
+ * <li>
+ * {@code constructors.declaration.grouping}
+ * </li>
+ * </ul>
+ *
+ * @since 10.16.0
+ */
+
+@FileStatefulCheck
+public class ConstructorsDeclarationGroupingCheck extends AbstractCheck {
```

Comment:
Can you make it stateless?
It will work well when we make checkstyle multithreading.

Reply:
Firstly I tried making it stateless using other ways but the `mvn clean verify` was giving errors so I have added the suppressions to make it work.

https://github.com/checkstyle/checkstyle/blob/fde8f4aba57f969543a0bf3b9a67896ead38763a/src/test/java/com/puppycrawl/tools/checkstyle/internal/ImmutabilityTest.java#L156-L157

https://github.com/checkstyle/checkstyle/blob/fde8f4aba57f969543a0bf3b9a67896ead38763a/src/test/java/com/puppycrawl/tools/checkstyle/internal/ImmutabilityTest.java#L193

After adding the suppressions, `mvn clean verify` was running perfectly

Reply:
> mutating them would make test cases fail.

@Zopsss how is that? if the tests fail when hardcoding the mutation then how they survive in the first place maybe you hardcode the mutation in the wrong way

Reply:
> how is that?

NVM it was showing that many surviving mutations because I didn't added the Test class in the TargetTests of pitest-coding-1. Thanks for helping me out. Now it is showing only 3 surviving mutations :)


@romani Looking into the surviving mutations now 🫡

Reply:
I am "unresolving" this item to continue discussion. We should be able to make this check stateless by using all class definitions as required tokens (CLASS_DEF_, ENUM_DEF, etc...), then checking class members. This should allow us to make this check stateless, perhaps at the cost of a slightly larger memory footprint. @Zopsss @romani thoughts?

Reply:
> We should be able to make this check stateless by using all class definitions as required tokens (CLASS_DEF_, ENUM_DEF, etc...), then checking class members.

do you mean we should use an approach like this?

https://github.com/checkstyle/checkstyle/blob/55f5c966bfa61b387110db6c4780f259a21676c8/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/OverloadMethodsDeclarationOrderCheck.java#L78-L92

The [OverloadMethodsDeclarationOrderCheck](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/OverloadMethodsDeclarationOrderCheck.java) uses this approach and it is stateless however the `visitToken()` gets called every time for class, enum, interface, etc... Then we're travelling the whole block to check that all the methods are grouped together. By this approach we can make the check stateless but then we will need to travel the whole class, enum, records, etc, we will be checking each and every tokens of the block, this is a time consuming method as we're travelling through the whole block and checking every tokens. 

We're currently calling the Check for only `CTOR` and `COMPACT_CTOR` tokens only, by using this implementation we're saving a lot of time instead of travelling the whole block. This is an optimized method to check that all constructors are grouped together or not.

We're also clearing the HashMap every time before calling the `visitToken()`, so the HashMap does not contain any unnecessary data:

https://github.com/checkstyle/checkstyle/blob/cadb0dc3a4635dd53eaa24ad6d41fd6e2cef7fd2/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ConstructorsDeclarationGroupingCheck.java#L83-L86 

If we make the current implementation stateless then we cannot use this optimized approach. The check will take more time to execute. So I guess if we're able to make the Check optimized then we should be using the current approach. The Check implementation will also become complex if we use the OverloadMethodsDeclarationOrder's approach. Correct me if I'm wrong.

Reply:
Sorry, I didn't read all this but just a few comments. We aren't multithreaded any time soon, so its not a big deal if we can't change the type of check there. Since we don't have an implementation of multithreading there is no telling how big a performance there may be. I don't see stateless being a big performance improvement. It just means we don't have to instantiate the check for each new file it processes. While new instances are cheap, it really depends on how we will do that instantiation and populate any property values.

Reply:
After reading @rnveach's comment, do we still need to transition to the stateless check or is it okay to keep this check stateful? @nrmancuso @romani please tell me what do you think about this?